src/Microsoft.ML.Data/Transforms/KeyToValueTransform.cs

```diff
@@ -67,386 +79,513 @@ private static VersionInfo GetVersionInfo()
-                    ctx.Pfa.AddFunc(funcName, new JArray(PfaUtils.Param("key", PfaUtils.Type.Int)),
-                        outType, PfaUtils.If(PfaUtils.Call("<", "key", 0), defaultToken,
-                        PfaUtils.Index(cellRef, "key")));
-                    var funcRef = PfaUtils.FuncRef("u." + funcName);
-                    return PfaUtils.Call("a.map", srcToken, funcRef);
+                    var outCol = (IColInput)toOutput[i];
+                    cols[i] = (inputNames[outCol.Input], outputNames[toOutput[i]]);
                 }
-                return PfaUtils.If(PfaUtils.Call("<", srcToken, 0), defaultToken, PfaUtils.Index(cellRef, srcToken));
+                return new KeyToValueEstimator(env, cols);
             }
         }
+
+    }
+
+    public sealed class KeyToValueEstimator : TrivialEstimator<KeyToValueTransform>
+    {
+        public KeyToValueEstimator(IHostEnvironment env, string columnName)
+            : base(Contracts.CheckRef(env, nameof(env)).Register(nameof(KeyToValueEstimator)), new KeyToValueTransform(env, columnName))
+        {
+        }
+
+        public KeyToValueEstimator(IHostEnvironment env, params (string input, string output)[] columns)
+            : base(Contracts.CheckRef(env, nameof(env)).Register(nameof(KeyToValueEstimator)), new KeyToValueTransform(env, columns))
+        {
+        }
+
+        public override SchemaShape GetOutputSchema(SchemaShape inputSchema)
+        {
+            Host.CheckValue(inputSchema, nameof(inputSchema));
+            var result = inputSchema.Columns.ToDictionary(x => x.Name);
+            foreach (var colInfo in Transformer.Columns)
+            {
+                if (!inputSchema.TryFindColumn(colInfo.input, out var col))
+                    throw Host.ExceptSchemaMismatch(nameof(inputSchema), "input", colInfo.input);
+                if (!col.IsKey)
+                    throw Host.ExceptSchemaMismatch(nameof(inputSchema), "input", colInfo.input, "key type", col.GetTypeString());
+
+                if (!col.Metadata.TryFindColumn(MetadataUtils.Kinds.KeyValues, out var keyMetaCol))
+                    throw Host.ExceptParam(nameof(inputSchema), $"Input column '{colInfo.input}' doesn't contain key values metadata");
+
+                SchemaShape metadata = null;
+                if (col.Metadata.TryFindColumn(MetadataUtils.Kinds.SlotNames, out var slotCol))
+                    metadata = new SchemaShape(new[] { slotCol });
+
+                result[colInfo.output] = new SchemaShape.Column(colInfo.output, col.Kind, keyMetaCol.ItemType, keyMetaCol.IsKey, metadata);
+            }
+
+            return new SchemaShape(result.Values);
+        }
+    }
+
+    /// <summary>
+    /// Extension methods for the static-pipeline over <see cref="PipelineColumn"/> objects.
+    /// </summary>
+    public static class KeyToValueStaticExtensions
+    {
+        public static Scalar<TValue> KeyToValue<TKey, TValue>(this Scalar<Key<TKey, TValue>> input)
```

Comment:
>KeyToValue [](start = 37, length = 10)

Just a thought, I feel like the word `Key` might be redundant, in context. I wonder if `ToValue` might be sufficient?